pkg/daemon/daemon.go

```diff
@@ -1103,33 +1109,50 @@ func (dn *Daemon) checkStateOnFirstRun() error {
 		return dn.triggerUpdateWithMachineConfig(state.currentConfig, state.desiredConfig)
 	}
 
-	// We've validated our state.  In the case where we had a pendingConfig,
-	// make that now currentConfig.  We update the node annotation, delete the
-	// state file, etc.
-	//
-	// However, it may be the case that desiredConfig changed while we
-	// were coming up, so we next look at that before uncordoning the node (so
-	// we don't uncordon and then immediately re-cordon)
+	// We've validated state. Now, ensure that node is in desired state
+	var inDesiredConfig bool
+	if inDesiredConfig, err = dn.updateConfigAndState(state); err != nil {
+		return err
+	}
+	if inDesiredConfig {
+		return nil
+	}
+
+	if dn.recorder != nil {
+		dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeNormal, "BootResync", fmt.Sprintf("Booting node %s, currentConfig %s, desiredConfig %s", dn.node.Name, state.currentConfig.GetName(), state.desiredConfig.GetName()))
+	}
+	// currentConfig != desiredConfig, and we're not booting up into the desiredConfig.
+	// Kick off an update.
+	return dn.triggerUpdateWithMachineConfig(state.currentConfig, state.desiredConfig)
+}
+
+// updateConfigAndState updates node to desired state, labels nodes as done and uncordon
+func (dn *Daemon) updateConfigAndState(state *stateAndConfigs) (bool, error) {
```

Comment:
im a little confused about why we are returning bool here when we seem to just care about whether we return an error or not? where this is called we are always first checking error , so im not clear on why we need to return the bool bc it seems to always come down to: is there an error or not (and there's only one case where err is nil)

Reply:
is there are case where you return false, nil? The "All good!" comment below implies no?

Reply:
There is one case where we get `false, nil`, if `inDesiredConfig := state.currentConfig.GetName() == state.desiredConfig.GetName()` evaluates as false, it skips the whole if block and just returns false.

Reply:
The naming is somewhat confusing, we probably should have named it `isInDesiredConfig` or something originally. But `inDesiredConfig` can be false

Reply:
could we delete that "All good!" comment below? since that's back from when it just returned nil?